### PR TITLE
SAAS-2975: Added EMR Desired Instance Type Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/emr/emrDesiredInstanceType.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            }
+
+            for (const cluster of listClusters.data) {
+                if (!cluster.Id) continue;
+
+                const resource = cluster.ClusterArn;
+
+                const listInstanceGroups = helpers.addSource(cache, source,
+                    ['emr', 'listInstanceGroups', region, cluster.Id]);
+                    // console.log(JSON.stringify(listInstanceGroups, null, 2));
```
Comment (sadeed12345) [inline plugins/aws/emr/emrDesiredInstanceType.js:65]:
```suggestion
```

File: plugins/aws/emr/emrDesiredInstanceType.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+        const emr_desired_master_instance_type = settings.emr_desired_master_instance_type || this.settings.emr_desired_master_instance_type.default;
+        const emr_desired_core_instance_type = settings.emr_desired_core_instance_type || this.settings.emr_desired_core_instance_type.default;
+
+        if (!emr_desired_master_instance_type.length && !emr_desired_core_instance_type.length) return callback(null, results, source);
+
+        async.each(regions.emr, function(region, rcb) {
+            const listClusters = helpers.addSource(cache, source,
+                ['emr', 'listClusters', region]);
+                // console.log(JSON.stringify(cache.emr, null ,2));
```
Comment (sadeed12345) [inline plugins/aws/emr/emrDesiredInstanceType.js:42]:
```suggestion
```

File: plugins/aws/emr/emrDesiredInstanceType.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        emr_desired_core_instance_type: {
+            name: 'EMR Desired Core Instance Type',
+            description: 'Comma-separated list of desired core instance types for EMR',
+            regex: '^.*$',
+            default: ''
+        }
+    },
+
+    run: function(cache, settings, callback) {  
+        // console.log(JSON.stringify(cache, null, 2));
```
Comment (sadeed12345) [inline plugins/aws/emr/emrDesiredInstanceType.js:29]:
```suggestion
```

File: collectors/aws/collector.js [hunk 4]
```diff
@@ -536,7 +536,7 @@ var calls = {
             paginate: 'Marker',
             params: {
                 ClusterStates: [
-                    'RUNNING'
+                    'RUNNING','TERMINATED_WITH_ERRORS', 'STARTING','TERMINATING'
```
Comment (AkhtarAmir) [inline collectors/aws/collector.js:539]: Let's discuss about this tomorrow

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: collectors/aws/collector.js (hunk 4 above)
Comment (sadeed12345) [inline collectors/aws/collector.js:539]:
```suggestion
                    'RUNNING'
```

**Review by AkhtarAmir — APPROVED** [review]

File: plugins/aws/emr/emrDesiredInstanceType.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'EMR Cluster Desired Instance Type',
+    category: 'EMR',
+    domain: 'Compute',
+    description: 'Ensure AWS Elastic MapReduce (EMR) clusters have the desired instance enabled.',
```
Comment (dipsubha06) [inline plugins/aws/emr/emrDesiredInstanceType.js:8]:
```suggestion
    description: 'Ensure AWS Elastic MapReduce (EMR) clusters are using desired instance type.
```

**Review by dipsubha06 — APPROVED** [review]

**Pull request merged**